### ISSUE TITLE
Support for Razer BlackWidow V4

### ISSUE THREAD

**Issue by JoeWhizer** [openrazer/razer-drivers#2163]
### Device Info

* Name: BlackWidow v4
* Product Number: RZ03-04692100-R3G1
* Model Number: RZ03-0469

### Device Information
```
Please post the output of
    lsusb -d 1532: -v

`Bus 005 Device 007: ID 1532:0287 Razer USA, Ltd Razer BlackWidow V4
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0287 Razer BlackWidow V4
  bcdDevice            1.01
  iManufacturer           1 Razer
  iProduct                2 Razer BlackWidow V4
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
Device Qualifier (for other device speed):
  bLength                10
  bDescriptorType         6
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  bNumConfigurations      0
Device Status:     0x0000
  (Bus Powered)
`
```

```
Please post the output of
    ls -lah /dev/input/by-id/

total 0
drwxr-xr-x 2 root root 300 14. Nov 16:48 .
drwxr-xr-x 4 root root 660 14. Nov 16:48 ..
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Burr-Brown_from_TI_USB_Audio_CODEC-event-if03 -> ../event8
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-©Microsoft_Corporation_Controller_0A5101D-event-joystick -> ../event5
lrwxrwxrwx 1 root root   6 14. Nov 16:48 usb-©Microsoft_Corporation_Controller_0A5101D-joystick -> ../js0
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-event-mouse -> ../event2
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-if01-event-kbd -> ../event12
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-if02-event-kbd -> ../event14
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-mouse -> ../mouse0
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-event-if01 -> ../event22
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-event-if04 -> ../event24
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-event-kbd -> ../event20
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-if01-event-kbd -> ../event21
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-if02-event-mouse -> ../event23
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-if02-mouse -> ../mouse1

```

```
Please post the output of
    usbhid-dump -m 1532 -ed

005:005:003:DESCRIPTOR         1699978231.860639
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 95 5A B1 01 C0

005:005:002:DESCRIPTOR         1699978231.863627
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

005:005:001:DESCRIPTOR         1699978231.866627
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 0E 81 00 C0 05 0C 09 01 A1 01 85
 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75 10 81
 00 75 08 95 0D 81 01 C0 05 01 09 80 A1 01 85 03
 19 81 29 83 15 00 25 01 75 01 95 03 81 02 95 05
 81 01 75 08 95 0E 81 01 C0 05 01 09 00 A1 01 85
 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95
 0F 81 00 C0 05 01 09 00 A1 01 85 05 09 03 15 00
 26 FF 00 35 00 46 FF 00 75 08 95 0F 81 00 C0

005:005:000:DESCRIPTOR         1699978231.869627
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 0B 81 03
 05 0C 0A 38 02 15 81 25 7F 75 08 95 01 81 06 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 C0

005:007:004:DESCRIPTOR         1699978231.871632
 05 59 09 01 A1 01 85 01 09 02 A1 02 09 03 15 00
 27 FF FF 00 00 75 10 95 01 B1 03 09 04 09 05 09
 06 09 07 09 08 15 00 27 FF FF FF 7F 75 20 95 05
 B1 03 C0 85 02 09 20 A1 02 09 21 15 00 27 FF FF
 00 00 75 10 95 01 B1 02 C0 85 03 09 22 A1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 01 B1 02 09 23
 09 24 09 25 09 27 09 26 15 00 27 FF FF FF 7F 75
 20 95 05 B1 02 09 28 09 29 09 2A 09 2C 09 2D 15
 00 26 FF 00 75 08 95 05 B1 02 C0 85 04 09 50 A1
 02 09 03 09 55 15 00 25 08 75 08 95 02 B1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 08 B1 02 09 51
 09 52 09 53 09 51 09 52 09 53 09 51 09 52 09 53
 09 51 09 52 09 53 09 51 09 52 09 53 09 51 09 52
 09 53 09 51 09 52 09 53 09 51 09 52 09 53 15 00
 26 FF 00 75 08 95 18 B1 02 C0 85 05 09 60 A1 02
 09 55 15 00 25 08 75 08 95 01 B1 02 09 61 09 62
 15 00 27 FF FF 00 00 75 10 95 02 B1 02 09 51 09
 52 09 53 15 00 26 FF 00 75 08 95 03 B1 02 C0 85
 06 09 70 A1 02 09 71 15 00 25 01 75 08 95 01 B1
 02 C0 85 07 06 00 FF 15 00 26 FF 00 09 02 75 08
 95 3F B1 02 C0 05 0C 09 01 A1 01 85 08 19 D0 29
 D7 15 00 25 01 75 01 95 08 81 02 C0

005:007:003:DESCRIPTOR         1699978231.874630
 05 01 09 00 A1 01 06 00 FF 09 03 15 00 25 01 75
 08 95 5A B1 01 C0

005:007:002:DESCRIPTOR         1699978231.877628
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 C0

005:007:001:DESCRIPTOR         1699978231.880631
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 29 A0 15 00 25 01
 75 01 95 A0 81 02 05 08 19 01 29 03 15 00 25 01
 75 01 95 03 91 02 95 05 91 01 C0 05 0C 09 01 A1
 01 85 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75
 10 81 00 75 08 95 13 81 01 C0 05 01 09 80 A1 01
 85 03 19 81 29 83 15 00 25 01 75 01 95 03 81 02
 95 05 81 01 75 08 95 14 81 01 C0 05 01 09 00 A1
 01 85 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75
 08 95 15 81 00 C0 05 01 09 00 A1 01 85 05 09 03
 15 00 26 FF 00 35 00 46 FF 00 75 08 95 15 81 00
 C0

005:007:000:DESCRIPTOR         1699978231.883631
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on **Razer Synapse** in a Windows Environment>

**Comment by JoeWhizer** [comment]
I will also take a look at 
https://github.com/openrazer/openrazer/blob/master/DEVELOPMENT.md

As v4 Pro already is supported, maybe I get it working myself - will spend some time on the weekend.

**Comment by kalzEOS** [comment]
I have this same keyboard and I'm getting a "failed" in front of the "Check device compatibilty" with:

- the following PIDs are not supported in the latest stable release of openrazer
- 1532:0287
- Check OpenRazer's issues and pull requests to see the status for your device

Everything else passes. Device is not recognized. I can provide more info if needed. Please let me know. 
Thanks

**Comment by JoeWhizer** [comment]
Wasn't successful yet, I've found the [commit](https://github.com/openrazer/openrazer/commit/1278b22b67e0288c58cda9a1eb1145ef3a3f6c85) for the v4 pro and did all the stuff related to v4 Pro for the v4, added the device ID also executed scripts for fakedriver generation and appstream file. All seems good, it compiled without problems.

I also could insmod the kernel driver without kernel panic, but the deamon  doesn't recognize the keyboard.

**Comment by z3ntu** [comment]
@JoeWhizer If you don't install the udev rules (and make them trigger) then the driver won't be rebound from hid-generic to razerkbd, so the device won't show up. You can also check `/sys/bus/hid/drivers/razerkbd/0*/` if there's stuff like `device_serial` in there, that would mean razerkbd is active.

**Comment by JoeWhizer** [comment]
there is device_serial after insmod the driver:
-r--r----- 1 root root 4,1k 25. Nov 14:08 /sys/bus/hid/drivers/razerkbd/0003:1532:0287.000E/device_serial

Also with the right usb id reference 1532:0287. 

but restarting the daemon with the loaded kernel module only find the mouse:
```
Nov 25 14:09:18 ArchDesk openrazer-daemon[99103]: 2023-11-25 14:09:18 | razer                          | INFO     | Initialising Daemon (v3.7.0). Pid: 99103
Nov 25 14:09:18 ArchDesk openrazer-daemon[99103]: 2023-11-25 14:09:18 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
Nov 25 14:09:18 ArchDesk openrazer-daemon[99103]: 2023-11-25 14:09:18 | razer                          | INFO     | Found device.0: 0003:1532:0099.0002
Nov 25 14:09:18 ArchDesk openrazer-daemon[99103]: 2023-11-25 14:09:18 | razer.device0                  | INFO     | Initialising device.0 RazerBasiliskV3
Nov 25 14:09:18 ArchDesk openrazer-daemon[99103]: 2023-11-25 14:09:18 | razer                          | INFO     | Serving DBus
```

**Comment by z3ntu** [comment]
Okay the driver side sounds good then. Are you sure then that the daemon you're running has support for your BlackWidow V4 with the correct PID there? When launching the daemon (with `--verbose` flag) then it should print all the devices it supports.

If you upload your branch somewhere I can also take a look.

**Comment by JoeWhizer** [comment]
Ahh, yeah you're right. The daemon doesn't output the v4, only v4 Pro

<details><summary>debug output</summary>

```
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAbyssus ---------------------------- (1532:0042)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAbyssus1800 ------------------------ (1532:0020)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAbyssus2000 ------------------------ (1532:005E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAbyssusEliteDVaEdition ------------- (1532:006A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAbyssusEssential ------------------- (1532:006B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAbyssusV2 -------------------------- (1532:005B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAnansi ----------------------------- (1532:010F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerAtherisReceiver -------------------- (1532:0062)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBaseStationChroma ------------------ (1532:0F08)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBaseStationV2Chroma ---------------- (1532:0F20)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasilisk --------------------------- (1532:0064)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskEssential ------------------ (1532:0065)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskUltimateReceiver ----------- (1532:0088)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskUltimateWired -------------- (1532:0086)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskV2 ------------------------- (1532:0085)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskV3 ------------------------- (1532:0099)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskV3ProWired ----------------- (1532:00AA)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskV3ProWireless -------------- (1532:00AB)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBasiliskXHyperSpeed ---------------- (1532:0083)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidow2019 --------------------- (1532:0241)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowChroma ------------------- (1532:0203)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowChromaOverwatch ---------- (1532:0211)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowChromaTournamentEdition -- (1532:0209)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowChromaV2 ----------------- (1532:0221)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowElite -------------------- (1532:0228)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowEssential ---------------- (1532:0237)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowLite --------------------- (1532:0235)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowStealth ------------------ (1532:011B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowStealthEdition ----------- (1532:010E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowTournamentEdition2014 ---- (1532:011C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowUltimate2012 ------------- (1532:010D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowUltimate2013 ------------- (1532:011A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowUltimate2016 ------------- (1532:0214)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowV3 ----------------------- (1532:024E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowV3MiniHyperspeed --------- (1532:0258)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowV3MiniHyperspeedWireless - (1532:0271)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowV3ProWired --------------- (1532:025A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowV3TK --------------------- (1532:0A24)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowV4Pro -------------------- (1532:028D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowXChroma ------------------ (1532:0216)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowXTournamentEditionChroma - (1532:021A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlackWidowXUltimate ---------------- (1532:0217)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade142021 ------------------------ (1532:0270)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade142022 ------------------------ (1532:028C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade152023 ------------------------ (1532:029E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade15Advanced2020 ---------------- (1532:0253)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade15Advanced2021 ---------------- (1532:0276)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade15AdvancedEarly2021 ----------- (1532:026D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade15AdvancedEarly2022 ----------- (1532:028A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade162023 ------------------------ (1532:029F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade172022 ------------------------ (1532:028B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade17Pro2021 --------------------- (1532:0279)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade17ProEarly2021 ---------------- (1532:026E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade182023 ------------------------ (1532:02A0)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade2018 -------------------------- (1532:0233)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade2018Base ---------------------- (1532:023B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade2018Mercury ------------------- (1532:0240)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade2019Adv ----------------------- (1532:023A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade2019Base ---------------------- (1532:0246)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBlade2019StudioEdition ------------- (1532:024D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeEarly2020Base ----------------- (1532:0255)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeEarly2021Base ----------------- (1532:026F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeEarly2022Base ----------------- (1532:027A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeLate2016 ---------------------- (1532:0224)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeMid2019Mercury ---------------- (1532:0245)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladePro2017 ----------------------- (1532:0225)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladePro2017FullHD ----------------- (1532:022F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladePro2019 ----------------------- (1532:0234)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeProEarly2020 ------------------ (1532:0256)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeProLate2016 ------------------- (1532:0210)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeProLate2019 ------------------- (1532:024C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeQHD --------------------------- (1532:020F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealth ----------------------- (1532:0205)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealth2019 ------------------- (1532:0239)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealthEarly2020 -------------- (1532:0252)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealthLate2016 --------------- (1532:0220)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealthLate2017 --------------- (1532:0232)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealthLate2019 --------------- (1532:024A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealthLate2020 --------------- (1532:0259)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBladeStealthMid2017 ---------------- (1532:022D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerBook2020 --------------------------- (1532:026A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerChargingPadChroma ------------------ (1532:0F26)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerChromaARGB ------------------------- (1532:0F1F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerChromaHDK -------------------------- (1532:0F09)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerChromaMugHolder -------------------- (1532:0F07)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerCobraPro --------------------------- (1532:00B0)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerCore ------------------------------- (1532:0215)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerCoreXChroma ------------------------ (1532:0F1A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerCynosaChroma ----------------------- (1532:022A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerCynosaChromaPro -------------------- (1532:022C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerCynosaLite ------------------------- (1532:023F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerCynosaV2 --------------------------- (1532:025E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdder1800 --------------------- (1532:0038)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdder2000 --------------------- (1532:004F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdder2013 --------------------- (1532:0037)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdder3500 --------------------- (1532:0054)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdder3_5G --------------------- (1532:0016)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdder3_5GBlack ---------------- (1532:0029)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderChroma ------------------- (1532:0043)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderElite -------------------- (1532:005C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderEssential ---------------- (1532:006E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderEssential2021 ------------ (1532:0098)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderEssentialWhiteEdition ---- (1532:0071)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV2 ----------------------- (1532:0084)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV2Lite ------------------- (1532:00A1)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV2Mini ------------------- (1532:008C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV2ProWired --------------- (1532:007C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV2ProWireless ------------ (1532:007D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV2XHyperSpeed ------------ (1532:009C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV3 ----------------------- (1532:00B2)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV3ProWired --------------- (1532:00B6)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathAdderV3ProWireless ------------ (1532:00B7)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerChroma ----------------- (1532:0204)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerEssential -------------- (1532:0118)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerExpert ----------------- (1532:0202)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerV2 --------------------- (1532:0295)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerV2ProTKLWired ---------- (1532:0298)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerV2ProTKLWireless ------- (1532:0296)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerV2ProWired ------------- (1532:0292)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDeathStalkerV2ProWireless ---------- (1532:0290)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerDiamondbackChroma ------------------ (1532:004C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerFirefly ---------------------------- (1532:0C00)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerFireflyHyperflux ------------------- (1532:0068)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerFireflyV2 -------------------------- (1532:0C04)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerGoliathus -------------------------- (1532:0C01)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerGoliathusExtended ------------------ (1532:0C02)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsman --------------------------- (1532:0227)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanElite ---------------------- (1532:0226)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanMini ----------------------- (1532:0257)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanMiniAnalog ----------------- (1532:0282)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanMiniJP --------------------- (1532:0269)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanTournamentEdition ---------- (1532:0243)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanV2 ------------------------- (1532:026C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanV2Analog ------------------- (1532:0266)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHuntsmanV2Tenkeyless --------------- (1532:026B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerHyperPollingWirelessDongle --------- (1532:00B3)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerImperator -------------------------- (1532:002F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerKraken71 --------------------------- (1532:0501)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerKraken71Alternate ------------------ (1532:0506)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerKraken71Chroma --------------------- (1532:0504)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerKraken71V2 ------------------------- (1532:0510)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerKrakenKittyEdition ----------------- (1532:0F19)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerKrakenUltimate --------------------- (1532:0527)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerLanceheadTE ------------------------ (1532:0060)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerLanceheadWired --------------------- (1532:0059)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerLanceheadWireless ------------------ (1532:005A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerLanceheadWirelessReceiver ---------- (1532:006F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerLanceheadWirelessWired ------------- (1532:0070)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerLaptopStandChroma ------------------ (1532:0F0D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerLaptopStandChromaV2 ---------------- (1532:0F2B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMamba2012Wired --------------------- (1532:0024)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMamba2012Wireless ------------------ (1532:0025)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMambaChromaWired ------------------- (1532:0044)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMambaChromaWireless ---------------- (1532:0045)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMambaElite ------------------------- (1532:006C)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMambaTE ---------------------------- (1532:0046)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMambaWirelessReceiver -------------- (1532:0072)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMambaWirelessWired ----------------- (1532:0073)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMouseBungeeV3Chroma ---------------- (1532:0F1D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerMouseDock -------------------------- (1532:007E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNaga2012 --------------------------- (1532:002E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNaga2014 --------------------------- (1532:0040)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaChroma ------------------------- (1532:0053)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaEpicChromaWired ---------------- (1532:003E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaEpicChromaWireless ------------- (1532:003F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaHex ---------------------------- (1532:0041)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaHexRed ------------------------- (1532:0036)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaHexV2 -------------------------- (1532:0050)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaLeftHanded2020 ----------------- (1532:008D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaProWired ----------------------- (1532:008F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaProWireless -------------------- (1532:0090)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaTrinity ------------------------ (1532:0067)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaV2HyperSpeedReceiver ----------- (1532:00B4)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNagaX ------------------------------ (1532:0096)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNommoChroma ------------------------ (1532:0517)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNommoPro --------------------------- (1532:0518)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerNostromo --------------------------- (1532:0111)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrbweaver -------------------------- (1532:0113)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrbweaverChroma -------------------- (1532:0207)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrnata ----------------------------- (1532:021F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrnataChroma ----------------------- (1532:021E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrnataV2 --------------------------- (1532:025D)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrnataV3X -------------------------- (1532:0294)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrnataV3X_Alternate ---------------- (1532:02A2)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrochi2011 ------------------------- (1532:0013)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrochi2013 ------------------------- (1532:0039)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrochiV2Bluetooth ------------------ (1532:0095)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrochiV2Receiver ------------------- (1532:0094)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOrochiWired ------------------------ (1532:0048)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerOuroboros -------------------------- (1532:0032)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerProClickMiniReceiver --------------- (1532:009A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerProClickReceiver ------------------- (1532:0077)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerProClickWired ---------------------- (1532:0080)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerRaptor27 --------------------------- (1532:0F12)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerTaipan ----------------------------- (1532:0034)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerTartarus --------------------------- (1532:0201)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerTartarusChroma --------------------- (1532:0208)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerTartarusV2 ------------------------- (1532:022B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerThunderbolt4DockChroma ------------- (1532:0F21)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViper ------------------------------ (1532:0078)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViper8KHz -------------------------- (1532:0091)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperMini -------------------------- (1532:008A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperMiniSEWired ------------------- (1532:009E)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperMiniSEWireless ---------------- (1532:009F)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperUltimateWired ----------------- (1532:007A)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperUltimateWireless -------------- (1532:007B)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperV2ProWired -------------------- (1532:00A5)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperV2ProWireless ----------------- (1532:00A6)
2023-11-25 16:15:12 | razer                          | DEBUG    | Loaded device specification: RazerViperV3HyperSpeed ------------------ (1532:00B8)
2023-11-25 16:15:12 | razer                          | INFO     | Found device.0: 0003:1532:0099.0002
2023-11-25 16:15:12 | razer.device0                  | INFO     | Initialising device.0 RazerBasiliskV3
```

</details>

here are all the changes I did:
https://github.com/JoeWhizer/openrazer/commit/9168e7b1685f4c5a0e5c579ca36abd7d217558d1

**Comment by z3ntu** [comment]
Not sure how you've installed the modified parts, but you could use e.g. this command to run the daemon from the source checkout: https://gist.github.com/z3ntu/c5ca8f6040d8934fafc5e686b234890e#file-gistfile1-txt-L12

**Comment by JoeWhizer** [comment]
Thanks! now at least I got something, but somehow it gets permission denied on device_serial and poll_rate and it throws an exception.

I ran this from source folder:
`PYTHONPATH="pylib:daemon" python3 ./daemon/run_openrazer_daemon.py -Fv --config=$PWD/daemon/resources/razer.conf`

With testdata its recognized and also listed in RazerGenie, but its not really making a connection to the device I suppose, only to the generated fake devices. So something probably is different from v4 Pro to v4 that causes this issues.

<details><summary>Debug output</summary>

```
2023-11-25 17:22:55 | razer                          | DEBUG    | Emitted Device Added Signal
2023-11-25 17:22:55 | razer                          | INFO     | Found valid device.1: 0003:1532:0287.001D
2023-11-25 17:22:55 | razer.device1                  | INFO     | Initialising device.1 RazerBlackWidowV4
2023-11-25 17:22:55 | razer.device1                  | WARNING  | getting serial: [Errno 13] Permission denied: '/sys/devices/pci0000:00/0000:00:08.1/0000:0d:00.3/usb5/5-3/5-3.2/5-3.2.2/5-3.2.2:1.2/0003:1532:0287.001D/device_serial'
2023-11-25 17:22:55 | razer.device1                  | DEBUG    | getting serial:  count:1
2023-11-25 17:22:55 | razer.device1                  | WARNING  | getting serial: [Errno 13] Permission denied: '/sys/devices/pci0000:00/0000:00:08.1/0000:0d:00.3/usb5/5-3/5-3.2/5-3.2.2/5-3.2.2:1.2/0003:1532:0287.001D/device_serial'
2023-11-25 17:22:55 | razer.device1                  | DEBUG    | getting serial:  count:2
2023-11-25 17:22:55 | razer.device1                  | WARNING  | getting serial: [Errno 13] Permission denied: '/sys/devices/pci0000:00/0000:00:08.1/0000:0d:00.3/usb5/5-3/5-3.2/5-3.2.2/5-3.2.2:1.2/0003:1532:0287.001D/device_serial'
2023-11-25 17:22:55 | razer.device1                  | DEBUG    | getting serial:  count:3
2023-11-25 17:22:55 | razer.device1                  | WARNING  | getting serial: [Errno 13] Permission denied: '/sys/devices/pci0000:00/0000:00:08.1/0000:0d:00.3/usb5/5-3/5-3.2/5-3.2.2/5-3.2.2:1.2/0003:1532:0287.001D/device_serial'
2023-11-25 17:22:55 | razer.device1                  | DEBUG    | getting serial:  count:4
2023-11-25 17:22:55 | razer.device1                  | WARNING  | getting serial: [Errno 13] Permission denied: '/sys/devices/pci0000:00/0000:00:08.1/0000:0d:00.3/usb5/5-3/5-3.2/5-3.2.2/5-3.2.2:1.2/0003:1532:0287.001D/device_serial'
2023-11-25 17:22:55 | razer.device1                  | DEBUG    | getting serial:  count:5

2023-11-25 17:22:55 | razer.device1                  | DEBUG    | DBus call set_poll_rate
Exception in thread Thread-2 (_collecting_udev_method):
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/daemon.py", line 605, in _collecting_udev_method
    self._add_device(d)
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/daemon.py", line 530, in _add_device
    razer_device = device_class(device_path=sys_path, device_number=device_number, config=self._config,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/hardware/keyboards.py", line 73, in __init__
    super().__init__(*args, **kwargs)
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/hardware/keyboards.py", line 25, in __init__
    super().__init__(*args, **kwargs)
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/hardware/device_base.py", line 1296, in __init__
    super().__init__(*args, **kwargs)
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/hardware/device_base.py", line 323, in __init__
    self.restore_dpi_poll_rate()
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/hardware/device_base.py", line 376, in restore_dpi_poll_rate
    poll_rate_func(self.poll_rate)
  File "/home/zalkortis/razerfork/daemon/openrazer_daemon/dbus_services/dbus_methods/mamba.py", line 319, in set_poll_rate
    with open(driver_path, 'w') as driver_file:
         ^^^^^^^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/sys/devices/pci0000:00/0000:00:08.1/0000:0d:00.3/usb5/5-3/5-3.2/5-3.2.2/5-3.2.2:1.2/0003:1532:0287.001D/poll_rate'
```

</details>

**Comment by z3ntu** [comment]
Then basically means the udev rules haven't run correctly since that should chown the files to plugdev

**Comment by Legion495** [comment]
Hello Everyone.

So I was wondering why my clicky V4 Pro would not work and it led me to this.
I am on Fedora. Now since I am very much clueless I will provide as much data as the OP here.

If someone would direct me with some steps I might be able to help.
Unless the pro variant is to be seen as a separate issue.

<details>
  <summary>lsusb</summary>

```
  legion495@fedora ~> lsusb -d 1532: -v

Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x028d [unknown]
  bcdDevice            1.01
  iManufacturer           1 Razer
  iProduct                2 Razer BlackWidow V4 Pro
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      88
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
Device Qualifier (for other device speed):
  bLength                10
  bDescriptorType         6
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  bNumConfigurations      0
Device Status:     0x0000
  (Bus Powered)

```

  </details>
  
  <details>
  <summary>ls -lah /dev/input/by-id</summary>

```
legion495@fedora ~> ls -lah /dev/input/by-id
total 0
drwxr-xr-x. 2 root root 280 23. Dez 17:42 ./
drwxr-xr-x. 4 root root 700 23. Dez 17:43 ../
lrwxrwxrwx. 1 root root  10 23. Dez 17:42 usb-Logitech_USB_Receiver-event-if01 -> ../event10
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Logitech_USB_Receiver-event-kbd -> ../event8
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Logitech_USB_Receiver-if01-event-mouse -> ../event9
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Logitech_USB_Receiver-if01-mouse -> ../mouse2
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Razer_Razer_BlackWidow_V4_Pro-event-if01 -> ../event5
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Razer_Razer_BlackWidow_V4_Pro-event-if04 -> ../event7
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Razer_Razer_BlackWidow_V4_Pro-event-kbd -> ../event2
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Razer_Razer_BlackWidow_V4_Pro-if01-event-kbd -> ../event3
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Razer_Razer_BlackWidow_V4_Pro-if02-event-mouse -> ../event6
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Razer_Razer_BlackWidow_V4_Pro-if02-mouse -> ../mouse1
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Valve_Software_Steam_Controller-event-mouse -> ../event4
lrwxrwxrwx. 1 root root   9 23. Dez 17:42 usb-Valve_Software_Steam_Controller-mouse -> ../mouse0

```

</details>
    
<details>
  <summary>usbhid-dump -m 1532 -ed</summary>


```
legion495@fedora ~> usbhid-dump -m 1532 -ed
001:004:004:DESCRIPTOR         1703350925.433243
 05 59 09 01 A1 01 85 01 09 02 A1 02 09 03 15 00
 27 FF FF 00 00 75 10 95 01 B1 03 09 04 09 05 09
 06 09 07 09 08 15 00 27 FF FF FF 7F 75 20 95 05
 B1 03 C0 85 02 09 20 A1 02 09 21 15 00 27 FF FF
 00 00 75 10 95 01 B1 02 C0 85 03 09 22 A1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 01 B1 02 09 23
 09 24 09 25 09 27 09 26 15 00 27 FF FF FF 7F 75
 20 95 05 B1 02 09 28 09 29 09 2A 09 2C 09 2D 15
 00 26 FF 00 75 08 95 05 B1 02 C0 85 04 09 50 A1
 02 09 03 09 55 15 00 25 08 75 08 95 02 B1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 08 B1 02 09 51
 09 52 09 53 09 51 09 52 09 53 09 51 09 52 09 53
 09 51 09 52 09 53 09 51 09 52 09 53 09 51 09 52
 09 53 09 51 09 52 09 53 09 51 09 52 09 53 15 00
 26 FF 00 75 08 95 18 B1 02 C0 85 05 09 60 A1 02
 09 55 15 00 25 08 75 08 95 01 B1 02 09 61 09 62
 15 00 27 FF FF 00 00 75 10 95 02 B1 02 09 51 09
 52 09 53 15 00 26 FF 00 75 08 95 03 B1 02 C0 85
 06 09 70 A1 02 09 71 15 00 25 01 75 08 95 01 B1
 02 C0 85 07 06 00 FF 15 00 26 FF 00 09 02 75 08
 95 3F B1 02 C0 05 0C 09 01 A1 01 85 08 19 D0 29
 D7 15 00 25 01 75 01 95 08 81 02 C0

001:004:003:DESCRIPTOR         1703350925.436237
 05 01 09 00 A1 01 06 00 FF 09 03 15 00 25 01 75
 08 95 5A B1 01 C0

001:004:002:DESCRIPTOR         1703350925.439242
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 75 08 95 01 81 03 05 0C 0A 38 02 15 81
 25 7F 75 08 95 01 81 06 05 01 09 38 15 81 25 7F
 75 08 95 01 81 06 09 30 09 31 16 00 80 26 FF 7F
 75 10 95 02 81 06 C0 C0

001:004:001:DESCRIPTOR         1703350925.442235
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 29 A0 15 00 25 01
 75 01 95 A0 81 02 05 08 19 01 29 03 15 00 25 01
 75 01 95 03 91 02 95 05 91 01 C0 05 0C 09 01 A1
 01 85 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75
 10 81 00 75 08 95 13 81 01 C0 05 01 09 80 A1 01
 85 03 19 81 29 83 15 00 25 01 75 01 95 03 81 02
 95 05 81 01 75 08 95 14 81 01 C0 05 01 09 00 A1
 01 85 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75
 08 95 15 81 00 C0 05 01 09 00 A1 01 85 05 09 03
 15 00 26 FF 00 35 00 46 FF 00 75 08 95 15 81 00
 C0

001:004:000:DESCRIPTOR         1703350925.445238
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

```

 </details>
  
  
  FYI my issue was solved by: "sudo dkms install openrazer-driver/3-7-0"

**Comment by z3ntu** [comment]
@Legion495 You have a different device, see #2044 for yours, and since that's part of OpenRazer v3.7.0 you should be able to update to that version and it should work. If not see https://github.com/openrazer/openrazer/wiki/Troubleshooting and if you need more help, join our community chat (see wiki) or open a new issue of type support.

**Comment by apj570** [comment]
Hello, just got this same keyboard and was wondering if you guys have come up with a solution.

**Comment by PlettserVladimir** [comment]
Hello! Thank you very much for your work! I am a happy owner of a keyboard Razer BlackWidow v4 (1532:0287) and also encountered the problem that the keyboard is not detected. I am ready to provide all possible assistance, I am currently on the Fedora 39 distribution.

**Comment by vividnightmare** [comment]
What is needed to progress this device to being supported?

**Comment by z3ntu** [comment]
@vividnightmare Somebody needs to add support for this device, https://github.com/openrazer/openrazer/blob/master/DEVELOPMENT.md has a rough outline of what needs to be done.

**Comment by xavierog** [comment]
@JoeWhizer Did you get any progress with this keyboard? Your latest message hints you were one permission/udev issue away from a breakthrough.

**Comment by xavierog** [comment]
FWIW, I have spawned a Debian Sid VM, cloned this repository, applied JoeWhizer's patch on top of it and my first tests (i.e. just playing with the various scripts in `examples/`) reflect it simply works. I am going to set up a desktop environment and polychromatic to ensure it works from end to end (including udev rules). Stay tuned.

**Comment by xavierog** [comment]
I think I clicked everything that could be clicked in Polychromatic and, overall, it works:

- DEVICEIMAGE is correct
- Device info looks correct
- brightness works
- built-in effects (and their options) work
- game mode works
- macros work, except on M6 (presumably just a small extra change)
- custom effects work; caveat: the "device graphic" that is shown is the Pro variant and thus:
  - shows too many LEDs
  - shows a round unnamed button instead of the M6 key
  ... but this is [specific to Polychromatic](https://github.com/polychromatic/polychromatic/tree/master/data/devicemaps), not openrazer

Note: all tests performed at 500 Hz poll rate.

Outside Polychromatic: according to `xev`:

- multimedia keys are mapped correctly: XF86AudioPrev, XF86AudioPlay, XF86AudioNext, XF86AudioMute, XF86AudioLowerVolume, XF86AudioRaiseVolume
- M1 is mapped to XF86Tools
- M2 is mapped to XF86Launch5
- M3 is mapped to XF86Launch6
- M4 is mapped to XF86Launch7
- M5 is mapped to XF86Launch8
- M6 is not mapped (presumably yet another small change)

About udev rules: yes, they seem to work.
The testing procedure was super-hackish though:
```shell
# Prerequisite: remove the DKMS-generated razerkbd.ko.xz.
# This is rubbish but way faster than fighting DKMS, compression, CRC32 & whatnot.
# Manually insert the new driver:
insmod /root/openrazer/driver/razerkbd.ko
# Trigger udev rules:
find /sys/ /proc/ /dev/ -group plugdev -ls # 6 entries
udevadm trigger -c add
find /sys/ /proc/ /dev/ -group plugdev -ls # 523 entries
# log in as regular user, start patched OpenRazer daemon:
cd openrazer
PYTHONPATH="pylib:daemon" python3 ./daemon/run_openrazer_daemon.py -Fv --config=$PWD/daemon/resources/razer.conf
# this is a good time to check for permission errors in the daemon's output
# run polychromatic:
polychromatic-controller
```

After a few lines of code:
- [x] ensure M6 gets mapped to XF86Launch9
- [x] add M6 to the list of macro-friendly keys

Still TODO:
- tell the Polychromatic project to add the devicemap for the non-pro variant

**Comment by xavierog** [comment]
FWIW, I have published DEB packages for Debian Sid with support for the Razer BlackWidow V4 keyboard:

[2024-04-21 edit: following the release of OpenRazer 3.8.0 and its landing in Debian Sid, the formerly mentioned repository is no longer relevant]

**Comment by vividnightmare** [comment]
Right.

I think the idea here is to confirm that the patches are functional, add proper support for the device, and get them added to the upstream repository.

I have been using the patches linked by @JoeWhizer for several weeks now and can confirm that they add the necessary backend support for the device.

Can we get a developer/maintainer to review?  Or can @JoeWhizer perhaps create a pull request so we can get support into the next release?

**Comment by xavierog** [comment]
@vividnightmare: there is already a pull request (which includes @JoeWhizer's work): #2211 
So we indeed need developers/maintainers to take a look at it. But a quick glance at recent pull requests reflects little to no activity for several weeks now.

**Comment by xavierog** [comment]
@z3ntu thanks for merging and releasing; 3.8.0 is now available in Debian Sid and I confirm it works flawlessly.
